File: 1382.balance-a-binary-search-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func balanceBST(root *TreeNode) *TreeNode {
    return build(inorder(root))
}

func inorder(root *TreeNode) []*TreeNode {
    var ordered []*TreeNode
    var traverse func(*TreeNode)
    
    traverse = func(root *TreeNode) {
        if root != nil {
            traverse(root.Left)
            ordered = append(ordered, root)
            traverse(root.Right)
        }
    }

    traverse(root)
    return ordered
}

func build(nodes []*TreeNode) *TreeNode {
    if len(nodes) == 0 {
        return nil
    }
    m := len(nodes) / 2
    node := nodes[m]
    node.Left = build(nodes[:m])
    node.Right = build(nodes[m + 1:])
    return node
}

File: 1382.balance-a-binary-search-tree/solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun balanceBST(root: TreeNode?): TreeNode? {
        val nums = mutableListOf<Int>()

        fun inorder(root: TreeNode?) {
            root?.let { node ->
                inorder(node.left)
                nums.add(node.`val`)
                inorder(node.right)
            }
        }

        fun build(list: List<Int>): TreeNode? {
            if (list.isEmpty()) {
                return null
            }
            val m = list.size / 2
            return TreeNode(list[m]).apply {
                left = build(list.subList(0, m))
                right = build(list.subList(m + 1, list.size))
            }
        }

        inorder(root)
        return build(nums)
    }
}